src/sentry/search/snuba/executors.py

```diff
@@ -472,6 +478,323 @@ def query(
+        #         converted_name = converted_filter[0]
+
+        return converted_filter, converted_name
+
+    def query(
+        self,
+        projects,
+        retention_window_start,
+        group_queryset,
+        environments,
+        sort_by,
+        limit,
+        cursor,
+        count_hits,
+        paginator_options,
+        search_filters,
+        date_from,
+        date_to,
+    ):
+
+        now = timezone.now()
+        end = None
+        end_params = filter(None, [date_to, get_search_filter(search_filters, "date", "<")])
+        if end_params:
+            end = min(end_params)
+
+        if not end:
+            end = now + ALLOWED_FUTURE_DELTA
+
+        retention_date = max(filter(None, [retention_window_start, now - timedelta(days=90)]))
+
+        start_params = [date_from, retention_date, get_search_filter(search_filters, "date", ">")]
+        start = max(filter(None, start_params))
+
+        end = max([retention_date, end])
+
+        if start == retention_date and end == retention_date:
+            return self.empty_result
+
+        if start >= end:
+            return self.empty_result
+
+        sort_field = self.sort_strategies[sort_by]
+        chunk_growth = options.get("snuba.search.chunk-growth-rate")
+        max_chunk_size = options.get("snuba.search.max-chunk-size")
+        chunk_limit = limit
+        offset = 0
+        num_chunks = 0
+        hits = None
+
+        paginator_results = self.empty_result
+        result_groups = []
+        result_group_ids = set()
+
+        max_time = options.get("snuba.search.max-total-chunk-time-seconds")
+        time_start = time.time()
+
+        while (time.time() - time_start) < max_time:
+            num_chunks += 1
+
+            chunk_limit = min(int(chunk_limit * chunk_growth), max_chunk_size)
+
+            snuba_groups, total = self.snuba_search(
+                start=start,
+                end=end,
+                project_ids=[p.id for p in projects],
+                environment_ids=environments and [environment.id for environment in environments],
+                sort_field=sort_field,
+                cursor=cursor,
+                limit=chunk_limit,
+                offset=offset,
+                search_filters=search_filters,
+            )
+            metrics.timing("snuba.search.num_snuba_results", len(snuba_groups))
+            count = len(snuba_groups)
+            more_results = count >= limit and (offset + limit) < total
+            offset += len(snuba_groups)
+
+            if not snuba_groups:
+                break
+
+            group_to_score = dict(snuba_groups)
+            for group_id in group_to_score:
+                if group_id in result_group_ids:
+                    continue
+                group_score = group_to_score[group_id]
+                result_group_ids.add(group_id)
+                result_groups.append((group_id, group_score))
+
+            if hits is None:
```

Comment:
This seems wrong, you are calculating the hits for every chunk.

Reply:
The first time it comes through, hits will be None. The second time, it will not be None and won't be re run for the new chunk

Reply:
Then please move the check outside of the method, it is confusing now. When you see a call to `calculate_hits` you expect hits to be calculated.

Reply:
It was in the loop because the instantiation of SequencePaginator was in the loop, and it uses hits as a kwarg. This was part of the original implementation, so it got carried through. I moved hits calculation to before the loop, so we can return an empty result set if hits is 0 and avoid doing the full snuba query.